### PR TITLE
HTTP ref support for artifacts

### PR DESCRIPTION
**Pull request by annirudh**
Adds support for HTTP refs. If checksumming, we get the HTTP response headers and map them to the artifact manifest entry fields as follows:

`digest`: `Etag` if it exists, otherwise we just use the URL
`size`: `Content-Length`, otherwise we leave it empty

### CODE REVIEW

File: wandb/artifacts.py [hunk 1]
```diff
@@ -1038,9 +1041,58 @@ def _extra_from_obj(obj):
...
+            for data in response.iter_content(chunk_size=16 * 1024):
+                file.write(data)
+        return path
+
+    def store_path(self, artifact, path, name=None, checksum=True, max_objects=None):
+        name = name or os.path.basename(path)
+        if not checksum:
+            return [ArtifactManifestEntry(name, path, digest=path)]
+
+        with self._session.get(path, stream=True) as response:
```
Comment (vanpelt) [inline wandb/artifacts.py:1081]: Nice, I wonder if there are cases where we would want to just do a `head` request here...

**Comment by vanpelt** [comment]
We should probably also support the digest header:  https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Digest#:~:text=The%20Digest%20response%20HTTP%20header,have%20multiple%20different%20digest%20values.

File: wandb/artifacts.py (hunk 1 above)
Comment (annirudh) [reply wandb/artifacts.py:1081]: With `stream=True` I think it behaves the same way as the requests library won't start pulling in the bytes until `response.content` is accessed, but I can make this an explicit `HEAD` request instead.

**Comment by annirudh** [comment]
> We should probably also support the digest header: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Digest#:~:text=The%20Digest%20response%20HTTP%20header,have%20multiple%20different%20digest%20values.

Good call. Will fix.

File: wandb/artifacts.py (hunk 1 above)
Comment (annirudh) [reply wandb/artifacts.py:1081]: I think it's better to use `head()`, will fix.

Comment (vanpelt) [reply wandb/artifacts.py:1081]: Yeah, this is more robust.  Not all endpoints will support a HEAD request, but there could be some that only return these more detailed headers on a HEAD.  For now lets just keep it and maybe add a comment, I would think most will have the headers in the GET.

**Pull request merged**